lib/sudo-exec/src/monitor.rs

```diff
@@ -69,49 +84,37 @@ impl MonitorRelay {
 
         match result {
             Err(err) => {
-                backchannel.send((&err).into())?;
-                Err(err)
+                backchannel.send((&err).into()).unwrap();
+                exit(1);
             }
-            Ok((signals, command_pid, command_pgrp, command, pty_follower)) => Ok(Self {
-                signal_handlers: signals,
-                command_pid,
-                command_pgrp,
-                command,
-                _pty_follower: pty_follower,
-                backchannel,
-            }),
+            Ok((event_handler, command_pid, command_pgrp, command, pty_follower)) => (
+                Self {
+                    command_pid,
+                    command_pgrp,
+                    command,
+                    _pty_follower: pty_follower,
+                    backchannel,
+                },
+                event_handler,
+            ),
         }
     }
 
-    pub(super) fn run(mut self) -> ! {
-        loop {
-            // First we check if the command is finished
-            // FIXME: This should be polled alongside the signal handlers instead.
-            if let Ok(Some(status)) = self.command.try_wait() {
-                self.backchannel.send(status.into()).ok();
-
-                exit(0);
-            }
-
-            // Then we check any pending signals that we received. Based on `mon_signal_cb`.
-            //
-            // Right now, we rely on the fact that `poll` can be interrupted by a signal so this
-            // call doesn't block forever. We are guaranteed to receive `SIGCHLD` when the
-            // command terminates meaning that the `wait_command` call above will succeed on the
-            // next iteration of this loop. We won't have to rely on this behavior once we
-            // integrate `wait_command` into the `poll` itself.
-            if let Ok(infos) = self.signal_handlers.poll() {
-                for info in infos {
-                    self.relay_signal(info);
-                }
-            }
-        }
+    pub(super) fn run(mut self, event_handler: &mut EventHandler<Self>) -> ! {
+        event_handler.event_loop(&mut self);
+        drop(self);
+        exit(0);
     }
 
-    fn relay_signal(&self, info: SignalInfo) {
+    fn relay_signal(&mut self, info: SignalInfo, event_handler: &mut EventHandler<Self>) {
         match info.signal() {
             // FIXME: check `mon_handle_sigchld`
```

Comment:
can this FIXME be removed at this point?

Reply:
not yet, we have to handle a bunch of controlling terminal things here.